[LibWebRTC][GStreamer] Support for video encoding using GStreamer encoders

philn opened the pull request:
#### d908463e14d12fda5a42a1d3fb9cbf61041fd261
<pre>
[LibWebRTC][GStreamer] Support for video encoding using GStreamer encoders
<a href="https://bugs.webkit.org/show_bug.cgi?id=302248">https://bugs.webkit.org/show_bug.cgi?id=302248</a>

Reviewed by Xabier Rodriguez-Calvar.

This patch rewrites the LibWebRTC GStreamer-based VideoEncoder using the WebCodecs backend. For now
we enable this for VP8 and H.264. Other codecs rely on the native LibWebRTC encoders. Along the way
the GStreamer video encoder gained support for variable framerate and annexB (mostly relevant for
H.264, H.265 and AV1).

Driving-by, colorspace utilities were moved from the Cocoa-specific VideoFrame implementation to a
new header file, LibWebRTCVideoFrameUtilities.h.

* Source/WebCore/platform/graphics/gstreamer/VideoEncoderGStreamer.cpp:
(WebCore::GStreamerVideoEncoder::create):
(WebCore::GStreamerVideoEncoder::setBitRateAllocation):
(WebCore::GStreamerInternalVideoEncoder::initialize):
(WebCore::GStreamerInternalVideoEncoder::setBitRateAllocation):
(WebCore::GStreamerInternalVideoEncoder::applyRates):
* Source/WebCore/platform/graphics/gstreamer/VideoEncoderGStreamer.h:
* Source/WebCore/platform/gstreamer/VideoEncoderPrivateGStreamer.cpp:
(annexBStreamFormatCapsFieldValue):
(videoEncoderSetEncoder):
(videoEncoderSetCodec):
(videoEncoderSetFrameRate):
(webkit_video_encoder_class_init):
* Source/WebCore/platform/gstreamer/VideoEncoderPrivateGStreamer.h:
(WebKitVideoEncoderBitRateAllocation::create): Deleted.
(WebKitVideoEncoderBitRateAllocation::setBitRate): Deleted.
(WebKitVideoEncoderBitRateAllocation::getBitRate const): Deleted.
(WebKitVideoEncoderBitRateAllocation::scalabilityMode const): Deleted.
(WebKitVideoEncoderBitRateAllocation::WebKitVideoEncoderBitRateAllocation): Deleted.
* Source/WebCore/platform/mediarecorder/MediaRecorderPrivateGStreamer.cpp:
(WebCore::MediaRecorderPrivateBackend::configureVideoEncoder):
* Source/WebCore/platform/mediastream/gstreamer/GStreamerVideoRTPPacketizer.cpp:
(WebCore::GStreamerVideoRTPPacketizer::create):
* Source/WebCore/platform/mediastream/libwebrtc/LibWebRTCVideoFrameUtilities.h: Copied from Source/WebCore/platform/mediastream/libwebrtc/VideoFrameLibWebRTC.cpp.
(WebCore::colorSpaceFromLibWebRTCVideoFrame):
(WebCore::videoRotationFromLibWebRTCVideoFrame):
* Source/WebCore/platform/mediastream/libwebrtc/VideoFrameLibWebRTC.cpp:
(WebCore::VideoFrameLibWebRTC::colorSpaceFromFrame): Deleted.
* Source/WebCore/platform/mediastream/libwebrtc/gstreamer/GStreamerVideoEncoderFactory.cpp:
(WebCore::GStreamerEncodedImageBuffer::create):
(WebCore::GStreamerEncodedImageBuffer::GStreamerEncodedImageBuffer):
(WebCore::GStreamerVideoEncoderFactory::Create):
(WebCore::GStreamerVideoEncoderFactory::GetSupportedFormats const):
(WebCore::GStreamerEncodedImageBuffer::getBuffer const): Deleted.
(WebCore::GStreamerEncodedImageBuffer::getVideoResolution const): Deleted.
(WebCore::LibWebRTCGStreamerVideoEncoder::LibWebRTCGStreamerVideoEncoder): Deleted.
(WebCore::LibWebRTCGStreamerVideoEncoder::pipeline): Deleted.
(WebCore::LibWebRTCGStreamerVideoEncoder::makeElement): Deleted.
(WebCore::LibWebRTCGStreamerVideoEncoder::returnFromFlowReturn): Deleted.
(WebCore::LibWebRTCGStreamerVideoEncoder::createEncoder): Deleted.
(WebCore::LibWebRTCGStreamerVideoEncoder::AddCodecIfSupported): Deleted.
(WebCore::LibWebRTCGStreamerVideoEncoder::Caps): Deleted.
(WebCore::LibWebRTCGStreamerVideoEncoder::ConfigureSupportedCodec): Deleted.
(WebCore::LibWebRTCGStreamerVideoEncoder::SetRestrictionCaps): Deleted.
(WebCore::LibWebRTCGStreamerVideoEncoderH264::LibWebRTCGStreamerVideoEncoderH264): Deleted.
* Source/WebCore/platform/mediastream/libwebrtc/gstreamer/GStreamerVideoFrameLibWebRTC.cpp:
(WebCore::convertLibWebRTCVideoFrameToGStreamerSample):
(WebCore::convertGStreamerSampleToLibWebRTCVideoFrame):
(WebCore::GStreamerVideoFrameLibWebRTC::ToI420):
* Source/WebCore/platform/mediastream/libwebrtc/gstreamer/GStreamerVideoFrameLibWebRTC.h:
(WebCore::GStreamerVideoFrameLibWebRTC::sample):
(WebCore::GStreamerVideoFrameLibWebRTC::GStreamerVideoFrameLibWebRTC):
(WebCore::GStreamerVideoFrameLibWebRTC::getSample const): Deleted.
* Source/WebCore/platform/mediastream/libwebrtc/gstreamer/RealtimeIncomingVideoSourceLibWebRTC.cpp:
(WebCore::RealtimeIncomingVideoSourceLibWebRTC::OnFrame):
* Source/WebCore/platform/mediastream/mac/RealtimeIncomingVideoSourceCocoa.mm:
(WebCore::RealtimeIncomingVideoSourceCocoa::toVideoFrame):

Canonical link: <a href="https://commits.webkit.org/302840@main">https://commits.webkit.org/302840@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4b75a6dc85fcad978df461558acbd83c4613719f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/130438 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/2709 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/41392 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/137856 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/82032 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/132309 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/2715 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/2601 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/99378 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/82032 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/133385 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/162/builds/1984 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/116803 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/80076 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/154/builds/1909 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/34933 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/81115 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/110462 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/136/builds/35438 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/140334 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/2499 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/161/builds/2285 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/107893 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/2543 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/113145 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/107794 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27427 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/1937 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/31588 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/55458 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/2569 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/65957 "Built successfully") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/157/builds/2387 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/163/builds/2590 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/2495 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

calvaris commented on Source/WebCore/platform/graphics/gstreamer/VideoEncoderGStreamer.h:
```diff
@@ -29,22 +29,26 @@
     GStreamerVideoEncoder(const Config&, DescriptionCallback&&, OutputCallback&&);
     ~GStreamerVideoEncoder();
 
-private:
```
Do you really need to make all these methods public?

philn replied:
Yes!

calvaris commented on Source/WebCore/platform/gstreamer/VideoEncoderPrivateGStreamer.cpp:
```diff
@@ -475,6 +500,35 @@ void videoEncoderSetBitRateAllocation(WebKitVideoEncoder* self, RefPtr<WebKitVid
+    int framerateNumerator, framerateDenominator;
+    gst_util_double_to_fraction(frameRate, &framerateNumerator, &framerateDenominator);
+
+    return;
```
?!

philn replied:
Oops, debug leftovers

calvaris commented on Source/WebCore/platform/mediastream/libwebrtc/LibWebRTCVideoFrameUtilities.h:
```diff
@@ -0,0 +1,197 @@
+    case webrtc::VideoRotation::kVideoRotation_270:
+        return VideoFrameRotation::Left;
+    };
+    RELEASE_ASSERT_NOT_REACHED();
```
I think just an ASSERT_NOT_REACHED would be enough. I don't think we need to crash the process in all cases for an unexpected value.

calvaris commented on Source/WebCore/platform/mediastream/libwebrtc/gstreamer/GStreamerVideoEncoderFactory.cpp:
```diff
@@ -61,151 +54,167 @@ namespace WebCore {
-
 public:
-    static webrtc::scoped_refptr<GStreamerEncodedImageBuffer> create(GRefPtr<GstSample>&& sample)
+    static webrtc::scoped_refptr<GStreamerEncodedImageBuffer> create(std::span<const uint8_t>&& data)
```
spans do not need to be passed as &&

calvaris commented on Source/WebCore/platform/mediastream/libwebrtc/gstreamer/GStreamerVideoEncoderFactory.cpp:
```diff
@@ -61,151 +54,167 @@ namespace WebCore {
-    ~GStreamerEncodedImageBuffer() = default;
-    GStreamerEncodedImageBuffer(GRefPtr<GstSample>&& sample)
-        : m_sample(sample)
+    GStreamerEncodedImageBuffer(std::span<const uint8_t>&& data)
```
ditto

philn committed:
[LibWebRTC][GStreamer] Support for video encoding using GStreamer encoders
https://bugs.webkit.org/show_bug.cgi?id=302248

Reviewed by Xabier Rodriguez-Calvar.

This patch rewrites the LibWebRTC GStreamer-based VideoEncoder using the WebCodecs backend. For now
we enable this for VP8 and H.264. Other codecs rely on the native LibWebRTC encoders. Along the way
the GStreamer video encoder gained support for variable framerate and annexB (mostly relevant for
H.264, H.265 and AV1).

Driving-by, colorspace utilities were moved from the Cocoa-specific VideoFrame implementation to a
new header file, LibWebRTCVideoFrameUtilities.h.

* Source/WebCore/platform/graphics/gstreamer/VideoEncoderGStreamer.cpp:
(WebCore::GStreamerVideoEncoder::create):
(WebCore::GStreamerVideoEncoder::setBitRateAllocation):
(WebCore::GStreamerInternalVideoEncoder::initialize):
(WebCore::GStreamerInternalVideoEncoder::setBitRateAllocation):
(WebCore::GStreamerInternalVideoEncoder::applyRates):
* Source/WebCore/platform/graphics/gstreamer/VideoEncoderGStreamer.h:
* Source/WebCore/platform/gstreamer/VideoEncoderPrivateGStreamer.cpp:
(annexBStreamFormatCapsFieldValue):
(videoEncoderSetEncoder):
(videoEncoderSetCodec):
(videoEncoderSetFrameRate):
(webkit_video_encoder_class_init):
* Source/WebCore/platform/gstreamer/VideoEncoderPrivateGStreamer.h:
(WebKitVideoEncoderBitRateAllocation::create): Deleted.
(WebKitVideoEncoderBitRateAllocation::setBitRate): Deleted.
(WebKitVideoEncoderBitRateAllocation::getBitRate const): Deleted.
(WebKitVideoEncoderBitRateAllocation::scalabilityMode const): Deleted.
(WebKitVideoEncoderBitRateAllocation::WebKitVideoEncoderBitRateAllocation): Deleted.
* Source/WebCore/platform/mediarecorder/MediaRecorderPrivateGStreamer.cpp:
(WebCore::MediaRecorderPrivateBackend::configureVideoEncoder):
* Source/WebCore/platform/mediastream/gstreamer/GStreamerVideoRTPPacketizer.cpp:
(WebCore::GStreamerVideoRTPPacketizer::create):
* Source/WebCore/platform/mediastream/libwebrtc/LibWebRTCVideoFrameUtilities.h: Copied from Source/WebCore/platform/mediastream/libwebrtc/VideoFrameLibWebRTC.cpp.
(WebCore::colorSpaceFromLibWebRTCVideoFrame):
(WebCore::videoRotationFromLibWebRTCVideoFrame):
* Source/WebCore/platform/mediastream/libwebrtc/VideoFrameLibWebRTC.cpp:
(WebCore::VideoFrameLibWebRTC::colorSpaceFromFrame): Deleted.
* Source/WebCore/platform/mediastream/libwebrtc/gstreamer/GStreamerVideoEncoderFactory.cpp:
(WebCore::GStreamerEncodedImageBuffer::create):
(WebCore::GStreamerEncodedImageBuffer::GStreamerEncodedImageBuffer):
(WebCore::GStreamerVideoEncoderFactory::Create):
(WebCore::GStreamerVideoEncoderFactory::GetSupportedFormats const):
(WebCore::GStreamerEncodedImageBuffer::getBuffer const): Deleted.
(WebCore::GStreamerEncodedImageBuffer::getVideoResolution const): Deleted.
(WebCore::LibWebRTCGStreamerVideoEncoder::LibWebRTCGStreamerVideoEncoder): Deleted.
(WebCore::LibWebRTCGStreamerVideoEncoder::pipeline): Deleted.
(WebCore::LibWebRTCGStreamerVideoEncoder::makeElement): Deleted.
(WebCore::LibWebRTCGStreamerVideoEncoder::returnFromFlowReturn): Deleted.
(WebCore::LibWebRTCGStreamerVideoEncoder::createEncoder): Deleted.
(WebCore::LibWebRTCGStreamerVideoEncoder::AddCodecIfSupported): Deleted.
(WebCore::LibWebRTCGStreamerVideoEncoder::Caps): Deleted.
(WebCore::LibWebRTCGStreamerVideoEncoder::ConfigureSupportedCodec): Deleted.
(WebCore::LibWebRTCGStreamerVideoEncoder::SetRestrictionCaps): Deleted.
(WebCore::LibWebRTCGStreamerVideoEncoderH264::LibWebRTCGStreamerVideoEncoderH264): Deleted.
* Source/WebCore/platform/mediastream/libwebrtc/gstreamer/GStreamerVideoFrameLibWebRTC.cpp:
(WebCore::convertLibWebRTCVideoFrameToGStreamerSample):
(WebCore::convertGStreamerSampleToLibWebRTCVideoFrame):
(WebCore::GStreamerVideoFrameLibWebRTC::ToI420):
* Source/WebCore/platform/mediastream/libwebrtc/gstreamer/GStreamerVideoFrameLibWebRTC.h:
(WebCore::GStreamerVideoFrameLibWebRTC::sample):
(WebCore::GStreamerVideoFrameLibWebRTC::GStreamerVideoFrameLibWebRTC):
(WebCore::GStreamerVideoFrameLibWebRTC::getSample const): Deleted.
* Source/WebCore/platform/mediastream/libwebrtc/gstreamer/RealtimeIncomingVideoSourceLibWebRTC.cpp:
(WebCore::RealtimeIncomingVideoSourceLibWebRTC::OnFrame):
* Source/WebCore/platform/mediastream/mac/RealtimeIncomingVideoSourceCocoa.mm:
(WebCore::RealtimeIncomingVideoSourceCocoa::toVideoFrame):

Canonical link: https://commits.webkit.org/302840@main
